### PR TITLE
Gluu roles

### PR DESCRIPTION
**Pull request by JinpaLhawang**
# gate-local.yml

`spring:
  oauth2:
    client:
      clientId: <CLIENT_ID>
      clientSecret: <CLIENT_SECRET>
      userAuthorizationUri: https://<GLUU_SERVER>/oxauth/seam/resource/restv1/oxauth/authorize # Used to get an authorization code
      accessTokenUri: https://<GLUU_SERVER>/oxauth/seam/resource/restv1/oxauth/token # Used to get an access token
      scope: openid,email,profile,user_name,memberOf
    resource:
      userInfoUri: https://gluu.oneplatform.build/oxauth/seam/resource/restv1/oxauth/userinfo # Used to the current user's profile
    userInfoMapping: # Used to map the userInfo response to our User
      email:  email
      firstname: name
      lastname: family_name
      username: uid
      roles: roles

server:
  ssl:
    enabled: true
    keyStore: <PATH/TO/KEYSTORE>
    keyStorePassword: <PASSWORD>
    keyAlias: <ALIAS>`

### CODE REVIEW

**Comment by dstengle** [comment]
Is this an upstream merge combined with a functional PR? Can we get a separate maintenance merge of the upstream to make this more readable and stand-alone?

@skorten Can you help with maintenance here? [CC: @dpeterka ]

**Comment by JinpaLhawang** [comment]
Yeah I have upstream set up, so PR in my fork went to kenzanlabs. I could use some help setting up a different merge situation.

From: David Stenglein
Reply-To: kenzanlabs/gate
Date: Monday, November 14, 2016 at 9:17 AM
To: kenzanlabs/gate
Cc: Nicholas Eden-Walker, Author
Subject: Re: [kenzanlabs/gate] Gluu roles (#1)


Is this an upstream merge combined with a functional PR? Can we get a separate maintenance merge of the upstream to make this more readable and stand-alone?

@skorten<https://github.com/skorten> Can you help with maintenance here? [CC: @dpeterka<https://github.com/dpeterka> ]

**Comment by JinpaLhawang** [comment]
Oh, and just noticed the files changed includes all the changes not yet merged. Need to sync.

From: David Stenglein
Reply-To: kenzanlabs/gate
Date: Monday, November 14, 2016 at 9:17 AM
To: kenzanlabs/gate
Cc: Nicholas Eden-Walker, Author
Subject: Re: [kenzanlabs/gate] Gluu roles (#1)


Is this an upstream merge combined with a functional PR? Can we get a separate maintenance merge of the upstream to make this more readable and stand-alone?

@skorten<https://github.com/skorten> Can you help with maintenance here? [CC: @dpeterka<https://github.com/dpeterka> ]

**Comment by JinpaLhawang** [comment]
My personal fork has been synced with kenzanlabs/gate. Files changed is now reflective of the changes I made. I discussed with @pansonm about whether to add relevant code block to SpinnakerUserInfoTokenServices#loadAuthentication or to a separate UserRolesProvider for Gluu/AD/LDAP. The UserRolesProvider interface does not match the intention of the change to be made here, since it uses an userId to fetch roles data, whereas we already have the roles data and we just need to transform it into an array. We did think creating a UserRolesProvider would be the best abstraction, however in trying to set it up I realized an if/else conditional would still need to be added to SpinnakerUserInfoTokenServices, to pass the roles instead of just an userId, so it seemed like the lowest effort and simplest would be to just include the transform in that if/else block.

File: gate-web/src/main/groovy/com/netflix/spinnaker/gate/security/oauth2/SpinnakerUserInfoTokenServices.groovy [hunk 1]
```diff
@@ -74,6 +74,14 @@ class SpinnakerUserInfoTokenServices implements ResourceServerTokenServices {
     def username = details[userInfoMapping.username] as String
     def roles = userRolesProvider.loadRoles(username)
```
Comment (JinpaLhawang) [inline gate-web/src/main/groovy/com/netflix/spinnaker/gate/security/oauth2/SpinnakerUserInfoTokenServices.groovy]: When no UserRolesProvider is specified, default bean is created which returns an empty array (see AuthConfig#defaultUserRolesProvider: https://github.com/kenzanlabs/gate/blob/master/gate-web/src/main/groovy/com/netflix/spinnaker/gate/security/AuthConfig.groovy#L48)

**Pull request closed without merging**